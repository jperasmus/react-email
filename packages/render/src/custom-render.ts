import { ConcurrentRoot } from "react-reconciler/constants";
import { renderer } from "./renderer";
import { Root } from "./renderer/nodes/root";
import { Options } from "./options";
import { convert } from "html-to-text";
import { plainTextSelectors } from "./plain-text-selectors";
import { pretty } from "./utils/pretty";

interface IncompleteFiber {
  tag: number;
  memoizedState: Record<string, unknown> | null;
  sibling: IncompleteFiber | null;
  child: IncompleteFiber | null;
}

/**
  * Checks recursively through React's fibers to make sure that all suspense
  * boundaries are complete. It uses the fact that, once they are complete,
  * they will have a `memoizedState` of null. It also checks for an element being
  * a Suspense boundary by looking for an internal React field in the fibers.
  *
  * There is currently no other way that can be used to wait for these 
  * Suspense boundaries to be done except for checking the fibers themselves.
  *
  * NOTE: this code is very fragile and should be tested very carefully
  * with each `react-reconciler` update. This is because React might change
  * these internal structures and cause this code to break.
  */
function checkIfAllSuspensesAreDone(fiber: IncompleteFiber) {
  if (fiber.tag === 13 && fiber.memoizedState !== null) {
    // Suspense boundary is still pending
    return false;
  }
  let child = fiber.child;
  while (child !== null) {
    if (!checkIfAllSuspensesAreDone(child)) {
      return false;
    }
    child = child.sibling;
  }
  return true;
}

const renderToMarkup = (element: React.ReactElement) => {
  return new Promise<string>((resolve, reject) => {
    const root = new Root();

    const container = renderer.createContainer(
      root,
      ConcurrentRoot,
      {},
      false,
      true,
      "",
      (err) => {
        reject(err);
      },
      {},
    ) as {
      current: IncompleteFiber;
    };

    renderer.updateContainer(element, container, null, () => {
      function checkCompletion() {
        if (checkIfAllSuspensesAreDone(container.current)) {
          resolve(root.html);
        } else {
          // We check again after the current stack for the event loop is finished,
          // meaning that Suspense will already have retried.
          setTimeout(checkCompletion, 0);
        }
      }

      checkCompletion();
    });
  });
}

const doctype =
  '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

export const render = async (element: React.ReactElement, options?: Options) => {
  const markup = await renderToMarkup(element);

  const document = `${doctype}${markup}`;

  if (options?.plainText) {
    return convert(document, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
};
