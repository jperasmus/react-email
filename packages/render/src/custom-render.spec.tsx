import { Suspense } from "react";
import usePromise from "react-promise-suspense";
import { render } from "./custom-render";
import { renderAsync } from "./render-async";
import { VercelInviteUserEmail } from "./utils/vercel-template";

describe("render()", () => {
  it("works with `style` and a `className`", async () => {
    const EmailComponent = (props: { username: string }) => {
      return (
        <div
          className="this-is-my-class-name"
          style={{ transitionTimingFunction: "ease" }}
        >
          Hello, {props.username}
        </div>
      );
    };
    const html = await render(<EmailComponent username="banana man" />);

    expect(html).toMatchSnapshot();
  });

  it("works with Suspense", async () => {
    const wait = () => {
      return new Promise<void>((resolve) => {
        setTimeout(() => {
          resolve();
        }, 500);
      });
    };

    const EmailComponent = (props: { username: string }) => {
      const username = usePromise(
        async (value) => {
          await wait();
          return value;
        },
        [props.username],
      );

      return (
        <div
          className="this-is-my-class-name"
          style={{ transitionTimingFunction: "ease" }}
        >
          Hello, {username}
        </div>
      );
    };

    const html = await render(
      <div>
        <Suspense>
          <EmailComponent username="banana woman" />
        </Suspense>
      </div>,
    );

    expect(html).toMatchSnapshot();
  });

  test("with a demo email template vs react-dom's SSR", async () => {
    const template = <VercelInviteUserEmail />;
    const html = await render(template, { pretty: false });
    const htmlFromReactDom = await renderAsync(template, { pretty: false });
    expect(html).toBe(
      htmlFromReactDom
        .replaceAll(/<!--.*?-->/g, (match) =>
          match.startsWith("<!--[if mso]>") ? match : "",
        )
        .replaceAll("&#x27;", "'")
        .replaceAll("&quot;", '"'),
    );
  });
});

