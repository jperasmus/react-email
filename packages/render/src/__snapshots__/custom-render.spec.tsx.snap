// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render() > works with \`style\` and a \`className\` 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><div class="this-is-my-class-name" style="transition-timing-function:ease">Hello, banana man</div>"`;

exports[`render() > works with Suspense 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><div><div class="this-is-my-class-name" style="transition-timing-function:ease">Hello, banana woman</div></div>"`;
